sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/StartingPointShardsFinder.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.kinesis;
+
+import com.amazonaws.services.kinesis.model.Shard;
+import com.amazonaws.services.kinesis.model.ShardIteratorType;
+import com.google.common.base.Function;
+import com.google.common.base.Predicate;
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.Sets;
+
+import java.io.Serializable;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is responsible for establishing the initial set of shards that existed at the given
+ * starting point.
+ */
+class StartingPointShardsFinder implements Serializable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(StartingPointShardsFinder.class);
+
+  /**
+   * <p>Finds all the shards at the given startingPoint. This method starts by gathering the oldest
```

Comment:
Am I understanding correctly that this is needed because Kinesis API does not have a method "list shards that existed at the given time", only "list all shards that ever existed"? In that case, what's the deal with the scalability of that: does the set of shards that ever existed grow unboundedly over the lifetime of a Kinesis stream?

Reply:
Also: what's the problem with simply starting with *all* shards? Won't the irrelevant ones filter themselves out automatically via other logic in KinesisIO?

Reply:
Kinesis actually returns all the shards that remain within the stream’s retention period. So when shard is closed it will still be listed for up to 7 days (the current max retention period) and then it will disappear from the list. Starting with all the shards should be possible but it would cause different problems. Now when we want to read the stream from it’s beginning (up to max retention period) we just need to find initial shards and just follow the splits and merges to read new shards. When we start reading with all the shards at once, handling such splits and merges could be trickier. There would be cases when shard’s successors are already handled by some other thread (when such successor shard was listed at the beginning) and cases where successors are not yet handled (due to shard split operation happening after the initial listing). Moreover, related shards might be spread across UnboundedSource splits which could further complicate solution.